protocols/tls/src/certificate.rs

```diff
@@ -0,0 +1,145 @@
+            writer
+                .next()
+                .write_bitvec_bytes(signature, signature.len() * 8);
+        })
+    });
+    let mut ext = rcgen::CustomExtension::from_oid_content(LIBP2P_OID, contents);
+    ext.set_criticality(true);
+    ext
+}
+
+fn gen_signed_keypair(keypair: &identity::Keypair) -> (rcgen::KeyPair, rcgen::CustomExtension) {
+    let temp_keypair = rcgen::KeyPair::generate(&LIBP2P_SIGNATURE_ALGORITHM)
+        .expect("we pass valid parameters, and assume we have enough memory and randomness; qed");
+    let mut signing_buf =
+        [0u8; LIBP2P_SIGNING_PREFIX_LENGTH + LIBP2P_SIGNATURE_ALGORITHM_PUBLIC_KEY_LENGTH];
+    let public = temp_keypair.public_key_raw();
+    assert_eq!(
+        public.len(),
+        LIBP2P_SIGNATURE_ALGORITHM_PUBLIC_KEY_LENGTH,
+        "ed25519 public keys are {} bytes",
+        LIBP2P_SIGNATURE_ALGORITHM_PUBLIC_KEY_LENGTH
+    );
+    signing_buf[..LIBP2P_SIGNING_PREFIX_LENGTH].copy_from_slice(&super::LIBP2P_SIGNING_PREFIX[..]);
+    signing_buf[LIBP2P_SIGNING_PREFIX_LENGTH..].copy_from_slice(public);
+    let signature = keypair.sign(&signing_buf).expect("signing failed");
+    (
+        temp_keypair,
+        encode_signed_key(keypair.public(), &signature),
+    )
+}
+
+/// Generates a self-signed TLS certificate that includes a libp2p-specific certificate extension
+/// containing the public key of the given keypair.
+pub(crate) fn make_cert(keypair: &identity::Keypair) -> rcgen::Certificate {
+    let mut params = rcgen::CertificateParams::new(vec![]);
+    let (cert_keypair, libp2p_extension) = gen_signed_keypair(keypair);
+    params.custom_extensions.push(libp2p_extension);
+    params.alg = &LIBP2P_SIGNATURE_ALGORITHM;
+    params.key_pair = Some(cert_keypair);
+    rcgen::Certificate::from_params(params)
+        .expect("certificate generation with valid params will succeed; qed")
+}
+
+/// Extracts the `PeerId` from a certificate’s libp2p extension. It is erroneous
+/// to call this unless the certificate is known to be a well-formed X.509
+/// certificate with a valid libp2p extension. The certificate verifiers in this
+/// crate validate check this.
+///
+/// If you get `Err` from this function, there is a bug somewhere. Either you
+/// called it without checking the preconditions, or there is a bug in this
+/// library or one of its dependencies.
+pub fn extract_peerid(certificate: &[u8]) -> Result<libp2p_core::PeerId, webpki::Error> {
+    let mut id = None;
+    let cb = &mut |oid: untrusted::Input<'_>, value, _, _| match oid.as_slice_less_safe() {
+        super::LIBP2P_OID_BYTES => {
+            if id.is_some() {
+                error!(
+                    "multiple libp2p extensions should have been detected \
+                     earlier; something is wrong"
+                );
+                id = Some(Err(webpki::Error::UnknownIssuer))
+            }
+            id = Some(match extract_libp2p_peerid(value) {
```

Comment:
My understanding of `cb` is that it would potentially be applied multiple times and you would like to treat multiple libp2p extensions as an error. If so, this line overrides the error assignment in line 110, so the overall result could still be ok if `extract_linp2p_peerid` succeeds. I can not tell if this is intentional or not but given the error assignment above it would seem to me it is not.

Reply:
This is indeed a bug, thanks!

Reply:
That said, this bug is not actually relevant, as we reject the certificate earlier if multiple libp2p extensions are encountered.  Any error from this function results in a libp2p-quic panic.